cmd/clusterctl/client/cluster/ca_injection.go

```diff
@@ -0,0 +1,212 @@
+
+	return nil
+}
+
+// getCACertificateFor returns the ca certificate from the secret referred by the
+// Certificate object. It reads the namespaced name of the Certificate from the
+// injection annotation of the passed object.
+func (c *certificateInjectionVerifier) getCACertificateFor(ctx context.Context, obj unstructured.Unstructured) (string, error) {
+	annotationValue, ok := obj.GetAnnotations()[certManagerCAAnnotation]
+	if !ok || annotationValue == "" {
+		return "", fmt.Errorf("getting value for injection annotation")
+	}
+
+	certificateObjKey := splitObjectKey(annotationValue)
+
+	certificate := &unstructured.Unstructured{}
+	certificate.SetKind("Certificate")
+	certificate.SetAPIVersion("cert-manager.io/v1")
+
+	if err := c.Client.Get(ctx, certificateObjKey, certificate); err != nil {
+		return "", errors.Wrapf(err, "getting certificate %s", certificateObjKey)
+	}
+
+	secretName, _, err := unstructured.NestedString(certificate.Object, "spec", "secretName")
+	if err != nil || secretName == "" {
+		return "", errors.Wrapf(err, "reading .spec.secretName name from certificate %s", certificateObjKey)
+	}
+
+	secretObjKey := client.ObjectKey{Namespace: certificate.GetNamespace(), Name: secretName}
+	certificateSecret := &corev1.Secret{}
+	if err := c.Client.Get(ctx, secretObjKey, certificateSecret); err != nil {
+		return "", errors.Wrapf(err, "getting secret %s", &certificateObjKey)
+	}
+
+	ca, ok := certificateSecret.Data["ca.crt"]
+	if !ok {
+		return "", errors.Errorf("data for \"ca.crt\" not found in secret %s", secretObjKey)
+	}
+
+	return string(ca), nil
+}
+
+// splitObjectKey splits the string by the name separator and returns it as client.ObjectKey.
+func splitObjectKey(nameStr string) client.ObjectKey {
+	splitPoint := strings.IndexRune(nameStr, types.Separator)
+	if splitPoint == -1 {
+		return client.ObjectKey{Name: nameStr}
+	}
+	return client.ObjectKey{Namespace: nameStr[:splitPoint], Name: nameStr[splitPoint+1:]}
+}
+
+// checkObject gets the desired CA certificate and compares it relevant field on
+// the object which is first read from kube-apiserver.
+func (c *certificateInjectionVerifier) checkObject(ctx context.Context, obj unstructured.Unstructured) error {
+	// get the CA certificate from the Certificate's secret which is referred in an
+	// annotation at the object.
+	ca, err := c.getCACertificateFor(ctx, obj)
+	if err != nil {
+		return err
+	}
+
+	// Build the object key from the passed object.
+	objKey := client.ObjectKey{Namespace: obj.GetNamespace(), Name: obj.GetName()}
+
+	// Get the object and assert the CA certificate.
+	switch obj.GetKind() {
+	case customResourceDefinitionKind:
+		crd := &apiextensionsv1.CustomResourceDefinition{}
+		if err := c.Client.Get(ctx, objKey, crd); err != nil {
+			return err
+		}
+
+		if crd.Spec.Conversion.Webhook == nil || crd.Spec.Conversion.Webhook.ClientConfig == nil || string(crd.Spec.Conversion.Webhook.ClientConfig.CABundle) != ca {
```

Comment:
Should we instead skip the validation if Conversion.Webhook.ClientConfig doesn't exist?

It's a strange config for a CRD to have the annotation and no client config, but I think it's not on us to fail in that case.

I think if the conversion webhook is configured and the caBundle supposed to be used ClientConfig has to be configured before the cert-manger becomes active